### ISSUE TITLE
[question] Invoking package creation for different patch versions of dependencies

### ISSUE THREAD

**Issue by foundry-markf** [conan-io/conan#9732]
The subject doesn't do the question justice, so will explain further here. There is a real example of this which I'm looking at right now, but this is simplified.

Suppose we have:
* a mid-graph package called M, which has a runtime requirement on P, where
* package P at version 5.x.y, which contains shared libraries.
* An application A consumes M.

In theory, due to our default package_ids using semver, I can specify different patch versions of P in A's recipe, without needing to make any changes to M.

Unfortunately, P has a breaking change in a patch release, so doing the above breaks in M. (In reality, I get a runtime symbol lookup error on Mac in a dylib, due to there being different symbols between what was built and was is used at runtime, as a symbol has been removed in the latest patch.)

What I believe I need is a build against M that is different for the different patch versions of P.

I first thought I could reach for changing `semver_mode()` to `full_version_mode()` in M's recipe in respect of P, so that I get different package_ids for different builds of M, and downstream in A, I can specify my desired requirements to match a package_id of M. That sounds like what I want.

But then I had to stop, and think, how do I tell `conan create` (or local workflow), that I want two builds of M, with different versions of P? i.e. how can I affect `self.requires()` in M's recipe from the command line?

It could be a recipe option, but that would then make the package_id change redundant. So that doesn't seem right.

Am I trying to do something that isn't recommended?
Or am I missing a workflow?

Thanks!

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @foundry-markf 

You  have the ``--require-override`` cli option since 1.39, that you can do command line overrides, for this kind of use case.
Please have a look, and let me know if this is what you are looking for.

**Comment by foundry-markf** [comment]
Thanks @memsharded. I had completely overlooked that in changelogs. We are also on Conan 1.28 for historical reasons (although I'd really love to upgrade; it just can't happen overnight).

I can do some tests, but probably _not_ an option "right now".

I am also a little hesitant, in that it's experimental (we try not to adopt those bits into our workflow until they're considered stable), and this wording in the [docs](https://docs.conan.io/en/latest/reference/commands/consumer/install.html#require-override):

> Note however this would not be a generally recommended practice for production, it would be better to actually update the conanfiles to explicitly reflect in code which specific versions upstream are being used.

and I would want to use it in production.

**Comment by foundry-markf** [comment]
One other mechanism I thought of was to have duplicate recipes, apart from the requirements, but that also feels wrong in terms of future maintenance.

**Comment by memsharded** [comment]
> I am also a little hesitant, in that it's experimental (we try not to adopt those bits into our workflow until they're considered stable), and this wording in the docs:

Well, this is a chicken and egg problem 😅 It was recently introduced, and there are chances that it might need some adjustments after being used in real world setups. So we need to mark it experimental to be able to fix and improve things, and other hand, we need people using it to get that feedback to stabilize things.

Another possibility, maybe more technically sound and scalable (but I am not sure, would need to know better the use case) is the one of maintaining lockfiles for the supported configurations:

Lets say that you want to maintain builds of M/1.0 -> P/5.0.1, (one binary "pid1"), and M/1.0 -> P/5.0.2 (another binary of M, "pid2"). I assume that you have a version range in M/1.0, in the form of ``P/[~5.0]`` for example. Then maybe, maintaining 2 lockfiles, one that locks 5.0.1 and the other that locks 5.0.2 could work. This approach would be more scalable if that happens for  more dependencies which could also bump the patch version, and every lockfile will represent a different set of upstream versions.

And yet another possibility that I have seen some other users implementing is just bumping the M version (manually or automatically), when new versions upstream are released. M would not use version ranges, and instead declare pinned dependencies, when M conanfile.py is edited to account for the latest one, then its version is also bumped. Is a bit more manual, but control is total, so some users like it.

**Comment by foundry-markf** [comment]
I agree on the chicken-and-egg problem. It's a shame we're not tracking more recent Conan releases yet.

The lock files suggestion sounds interesting. We've discussed internally about how to use lockfiles for both end-products, but also how it might fit into thirdparty library builds. I hadn't considered this particular usage for it. But something I can give it a spin. Agree it would be more scalable if there were multiple dependencies that need to be managed. It does seem the Conan way of being exact for dependency versions.

In real life, the M package is PySide (P being Qt, and us having custom changes in one older version of Qt), so I'm less inclined to increment version numbers as it won't track what the library authors meant. I have considered massaging the version numbers to encode something about the dependency version (e.g. PySide/5.12.6-qt5.12.11), of which something similar occurs in generated shared libraries, but this feels more of a one-off workaround, rather than a general solution, as it doesn't scale at all well.

You've given me food for thought though, so I need to do a few experiments. Thank you!

**Comment by AbrilRBS** [comment]
Closing as Conan 1-specific.

Conan 2 offers a few way to override dependencies from profiles if the need arises and no modification of the recipe is desired
